redbot/core/commands/cog.py

```diff
@@ -0,0 +1,75 @@
+import inspect
+from typing import Callable, TYPE_CHECKING, Dict, List, Union, Awaitable, Optional
+
+import discord
+import discord.ext.commands
+
+from .commands import CogCommandMixin, CogGroupMixin, Command
+from ..i18n import Translator
+
+if TYPE_CHECKING:
+    from .context import Context
+
+
+__all__ = ["CogMeta", "Cog"]
+
+_ = Translator(__file__)
+
+
+class CogMeta(CogCommandMixin, CogGroupMixin, type):
+    """Metaclass for cog base class."""
+
+    checks: List[Callable[["Context"], Union[bool, Awaitable[bool]]]]
+    __translator: Optional[Callable[[str], str]]
+
+    def __init__(cls, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+        cls.checks = getattr(cls, "__commands_checks__", [])
+
+    @property
+    def all_commands(cls) -> Dict[str, Command]:
+        return {cmd.name: cmd for _, cmd in inspect.getmembers(cls) if isinstance(cmd, Command)}
+
+    @property
+    def qualified_name(cls) -> str:
+        return cls.__name__
+
+    async def can_run(cls, ctx: "Context") -> bool:
+        """Check if the given context passes this cog class's requirements.
+
+        Raises
+        ------
+        CommandError
+            Any exception which are raised from underlying checks.
+
+        """
+        ret = await discord.utils.async_all((pred(ctx) for pred in cls.checks))
+        if ret is False:
+            return ret
+        return await cls.requires.verify(ctx)
+
+    @classmethod
+    async def convert(mcs, ctx: "Context", argument: str) -> "Cog":
+        cog_instance = ctx.bot.get_cog(argument)
+        if cog_instance is None:
+            raise discord.ext.commands.BadArgument(_('Cog "argument" not found.'))
+        return type(cog_instance)
+
+    def translate(cls, untranslated: str) -> str:
+        if cls.__translator is not None:
+            return cls.__translator(untranslated)
+
+
+class Cog(metaclass=CogMeta):
```

Comment:
I'd prefer if the stuff in this were one more level of abstracted to better allow metaclass compatability.

This would entail chainging this to:

```py
class CogBase:
    """Base class for a cog."""
    def __init_subclass__(cls, **kwargs) -> None:
        cls.__translator = kwargs.pop("translator", None)
        super().__init_subclass__(**kwargs)

    @classmethod
    async def convert(cls, ctx: "Context", argument: str) -> "Cog":
        cog_instance = ctx.bot.get_cog(argument)
        if cog_instance is None:
            raise discord.ext.commands.BadArgument(_('Cog "argument" not found.'))
        return cog_instance

class Cog(CogBase, metaclass=CogMeta):
    pass
```

Doing this could allow for cooperative metaclassing in the following manner:

```py
# .abcs
from redbot.core import commands

class _MyMeta(type):
    ...

class MyMeta(_MyMeta, commands.CogMeta):
    pass

class CogBase(commands.CogBase, metaclass=MyMeta):
    pass
```
While I understand this is an edge case, it's reasonably accomodated, and avoids things like nearly completely breaking cogs using metaclasses to define APIs for static typing tools like mypy (see python/mypy#1996 )